src/metrics/summary.rs

```diff
@@ -0,0 +1,164 @@
+//! Module implementing an Open Metrics summary.
+//!
+//! See [`Summary`] for details.
+
+use crate::encoding::{EncodeMetric, MetricEncoder};
+
+use super::{MetricType, TypedMetric};
+use parking_lot::RwLock;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use quantiles::ckms::CKMS;
+
+/// Open Metrics [`Summary`] to measure distributions of discrete events.
+#[derive(Debug)]
+pub struct Summary {
+    target_quantile: Vec<f64>,
+    target_error: f64,
+    max_age_buckets: u64,
+    max_age_seconds: u64,
+    stream_duration: Duration,
+    inner: Arc<RwLock<InnerSummary>>,
+}
+
+impl Clone for Summary {
+    fn clone(&self) -> Self {
+        Summary {
+            target_quantile: self.target_quantile.clone(),
+            target_error: self.target_error,
+            max_age_buckets: self.max_age_buckets,
+            max_age_seconds: self.max_age_seconds,
+            stream_duration: self.stream_duration,
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct InnerSummary {
+    sum: f64,
+    count: u64,
+    quantile_streams: Vec<CKMS<f64>>,
+    // head_stream is like a cursor which carries the index
+    // of the stream in the quantile_streams that we want to query.
+    head_stream_idx: u64,
+    // timestamp at which the head_stream_idx was last rotated.
+    last_rotated_timestamp: Instant,
+}
+
+impl Summary {
+    /// Create a new [`Summary`].
+    pub fn new(max_age_buckets: u64, max_age_seconds: u64, target_quantile: Vec<f64>, target_error: f64) -> Self {
+        let mut streams: Vec<CKMS<f64>> = Vec::new();
+        for _ in 0..max_age_buckets {
+            streams.push(CKMS::new(target_error));
+        }
+
+        let stream_duration = Duration::from_secs(max_age_seconds / max_age_buckets);
+        let last_rotated_timestamp = Instant::now();
+
+        if target_quantile.iter().any(|&x| x > 1.0 || x < 0.0) {
+            panic!("Quantile value out of range");
+        }
+
+        Summary{
+            max_age_buckets,
+            max_age_seconds,
+            stream_duration,
+            target_quantile,
+            target_error,
+            inner: Arc::new(RwLock::new(InnerSummary {
+                sum: Default::default(),
+                count: Default::default(),
+                quantile_streams: streams,
+                head_stream_idx: 0,
+                last_rotated_timestamp,
+            }))
+        }
+    }
+
+    /// Observe the given value.
+    pub fn observe(&self, v: f64) {
+        self.rotate_buckets();
+
+        let mut inner = self.inner.write();
+        inner.sum += v;
+        inner.count += 1;
+
+        // insert quantiles into all streams/buckets.
+        for stream in inner.quantile_streams.iter_mut() {
+            stream.insert(v);
+        }
+    }
+
+    /// Retrieve the values of the summary metric.
+    pub fn get(&self) -> (f64, u64, Vec<(f64, f64)>) {
```

Comment:
```suggestion
    pub(crate) fn get(&self) -> (f64, u64, Vec<(f64, f64)>) {
```

Is this needed outside of the encode module?